src/fromager/vendor_rust.py

```diff
@@ -0,0 +1,192 @@
+        for filename in files:
+            content.append(root / filename)
+
+    content.sort()
+
+    for fn in content:
+        tar.add(
+            fn,
+            filter=_tar_reset,
+            recursive=False,
+            arcname=fn.relative_to(basedir),
+        )
+
+
+def _cargo_vendor(
+    manifests: typing.List[pathlib.Path],
+    project_dir: pathlib.Path,
+) -> None:
+    """Run cargo vendor"""
+    args = ["cargo", "vendor", f"--manifest-path={manifests[0]}"]
+    for manifest in manifests[1:]:
+        args.append(f"--sync={manifest}")
+    args.append(project_dir / VENDOR_DIR)
+    subprocess.check_call(args, stdout=subprocess.DEVNULL, env={})
+
+    # Remove large, unused archive and lib files. They are only
+    # used on Windows, macOS, and iOS. This makes the vendor bundle up to
+    # 60% smaller.
+    for pattern in VENDOR_FILTER_PATTERNS:
+        for filename in project_dir.glob(pattern):
+            filename.unlink()
+            filename.touch()  # create empty file
+
+
+def _fixup_cargo_manifest(
+    fileame: pathlib.Path,
+    *,
+    msrv: typing.Optional[Version] = None,
+):
+    """Modify Cargo.toml
+
+    pydantic-core has an unnecessary high MSRV.
+    """
+    with open(fileame, "r", encoding="utf-8") as f:
+        cfg = toml.load(f)
+
+    modified = False
+    if msrv is not None:
+        packaging = cfg.get("packaging", {})
+        rust_version = packaging.get("rust-version")
+        if rust_version is not None and Version(rust_version) >= msrv:
+            packaging["rust-version"] = str(msrv)
+            modified = True
+
+    if modified:
+        with open(fileame, "w", encoding="utf-8") as f:
+            toml.dump(cfg, f)
+
+
+def _cargo_config(project_dir: pathlib.Path):
+    """create .cargo/config.toml"""
+    dotcargo = project_dir / ".cargo"
+    config_toml = dotcargo / "config.toml"
+    try:
+        with open(config_toml, "r", encoding="utf-8") as f:
+            cfg = toml.load(f)
+    except FileNotFoundError:
+        dotcargo.mkdir(exist_ok=True)
+        cfg = {}
+
+    cfg.update(CARGO_CONFIG)
+
+    with open(config_toml, "w", encoding="utf-8") as f:
+        toml.dump(cfg, f)
+
+
+def vendor_rust(
+    sdist: pathlib.Path,
+    vendored: pathlib.Path,
+    *,
+    msrv: typing.Optional[Version] = RHEL9_MSRV,
+) -> None:
+    """Vendor Rust crates into a Python source distribution
+
+    Raises:
+      - NoCargoToml: if the sdist does not have a Cargo.toml file
+    """
+    if not sdist.suffixes != [".tar", ".gz"]:
+        raise ValueError(f"Invalid {sdist}, expected a tar.gz file")
+
+    with tempfile.TemporaryDirectory() as tmp:
+        tmpdir = pathlib.Path(tmp)
+        with tarfile.open(sdist) as tar:
+            # scan for Cargo.toml before extracting files
+            for filename in tar.getnames():
+                if os.path.basename(filename) == "Cargo.toml":
+                    break
+            else:
+                raise NoCargoToml(sdist)
```

Comment:
Maybe instead of an exception, just return a flag indicating that no work was done?

Reply:
The function now returns true or false instead of raising an exception.